src/store/io.rs

```diff
@@ -0,0 +1,166 @@
+use super::Store;
+use crate::node_pages_map::{Page, PAGE_SIZE};
+use crossbeam_channel::{Receiver, Sender, TryRecvError};
+use io_uring::{cqueue, opcode, squeue, types, IoUring};
+use slab::Slab;
+use std::{os::fd::AsRawFd, path::PathBuf, sync::Arc};
+
+const RING_CAPACITY: u32 = 64;
+
+// max number of inflight requests is bounded by the slab.
+const MAX_IN_FLIGHT: usize = 64;
+
+pub type HandleIndex = usize;
+
+#[derive(Clone, Copy)]
+pub enum IoKind {
+    Read,
+    Write,
+}
+
+pub struct IoCommand {
+    pub kind: IoKind,
+    pub handle: HandleIndex,
+    pub page_id: u64,
+    pub buf: Box<Page>,
+}
+
+pub struct CompleteIo {
+    command: IoCommand,
+    result: std::io::Result<()>,
+}
+
+/// Create an I/O worker managing an io_uring and sending responses back via channels to a number
+/// of handles.
+pub fn start_io_worker(
+    store: Arc<Store>,
+    num_handles: usize,
+) -> (Sender<IoCommand>, Vec<Receiver<CompleteIo>>) {
+    // main bound is from the pending slab.
+    let (command_tx, command_rx) = crossbeam_channel::bounded(1);
+    let (handle_txs, handle_rxs) = (0..num_handles)
+        .map(|_| crossbeam_channel::bounded(32))
+        .unzip();
+    std::thread::spawn(move || run_worker(store, command_rx, handle_txs));
+    (command_tx, handle_rxs)
+}
+
+fn run_worker(
+    store: Arc<Store>,
+    command_rx: Receiver<IoCommand>,
+    handle_tx: Vec<Sender<CompleteIo>>,
+) {
+    let mut pending: Slab<IoCommand> = Slab::with_capacity(MAX_IN_FLIGHT);
+
+    let mut ring = IoUring::<squeue::Entry, cqueue::Entry>::builder()
+        .setup_single_issuer()
+        .setup_sqpoll(100)
+        // .setup_iopoll
+        .build(RING_CAPACITY)
+        .expect("Error building io_uring");
+
+    loop {
+        // note: dropping the queues at the end of loop iteration performs `sync` implicitly
+        let (submitter, mut submit_queue, mut complete_queue) = ring.split();
+
+        // 1. process completions.
+        if !pending.is_empty() {
+            // block on next completion if at capacity.
+            if pending.len() == MAX_IN_FLIGHT && complete_queue.is_empty() {
+                // hack: this is not exposed from the io_uring or libc libraries.
+                const IORING_ENTER_GETEVENTS: u32 = 1;
+
+                // we just get events here, not submit.
+                // TODO: handle error
+                unsafe {
+                    submitter
+                        .enter::<libc::sigset_t>(
+                            0, // submit
+                            1, // complete
+                            IORING_ENTER_GETEVENTS,
```

Comment:
I really like this solution to wait for something to be completed! I didn't know it while I was working on the first iteration!